curvefs/src/client/warmup/warmup_manager.cpp

```diff
@@ -719,6 +732,74 @@ void WarmupManager::CollectMetrics(InterfaceMetric *interface, int count,
     interface->latency << (butil::cpuwide_time_us() - start);
 }
 
+bool WarmupManagerS3Impl::GetInodeSubPathParent(
+    fuse_ino_t inode, std::vector<std::string> subPath, fuse_ino_t *ret) {
+    if (subPath.empty()) {
+        return false;
+    }
+
+    if (subPath.size() == 1) {
+        *ret = inode;
+        return true;
+    }
+
+    fuse_ino_t nextInode;
+    std::string currentPath = subPath.front();
+    if (currentPath == "..") {
+        std::shared_ptr<InodeWrapper> inodeWrapper;
+        CURVEFS_ERROR statusCode = inodeManager_->GetInode(inode, inodeWrapper);
+        if (statusCode != CURVEFS_ERROR::OK) {
+            LOG(ERROR) << "get inode fail, inodeid = " << inode;
+            return false;
+        }
+        curve::common::UniqueLock lck = inodeWrapper->GetUniqueLock();
+        auto parents = inodeWrapper->GetParentLocked();
+        if (parents.empty()) {
+            // out of curvefs
+            // For example directory /A is symlinked to ../B
+            LOG(ERROR) << fmt::format(
+                "out of curvefs, inodeid = {}, subpath = {}", inode,
+                fmt::join(subPath, "/"));
+            return false;
+        }
+        // only support one parent
+        nextInode = parents[0];
+    } else if (currentPath == ".") {
+        nextInode = inode;
+    } else {
+        Dentry dentry;
+        CURVEFS_ERROR statusCode =
+            dentryManager_->GetDentry(inode, currentPath, &dentry);
+        if (statusCode != CURVEFS_ERROR::OK) {
+            LOG(ERROR) << "get dentry fail, inodeid = " << inode
+                       << ", path = " << currentPath;
+            return false;
+        }
+        if (dentry.type() == FsFileType::TYPE_SYM_LINK) {
+            // sym link
+            std::string symLink;
+            CURVEFS_ERROR statusCode =
+                fuseOpReadLink_(0, dentry.inodeid(), &symLink);
+            if (statusCode != CURVEFS_ERROR::OK) {
+                LOG(ERROR) << "readlink fail, inodeid = " << inode
+                           << ", path = " << currentPath;
+                return false;
+            }
+            std::vector<std::string> splitSymLink;
+            curve::common::AddSplitStringToResult(symLink, "/", &splitSymLink);
+            subPath.insert(subPath.begin() + 1, splitSymLink.begin(),
+                           splitSymLink.end());
+            nextInode = inode;
+        } else {
+            nextInode = dentry.inodeid();
```

Comment:
It's better to use iterative approach instead of recursion.

Reply:
> It's better to use iterative approach instead of recursion.

Here is to analyze the soft link, I don't think there will be too deep level calls. However, this can be modified. if you insist